text/0000-macro-naming.md

```diff
@@ -0,0 +1,165 @@
+
+Macro expansion order is also not defined by source order. E.g., in `foo!(); bar!();`,
+`bar` may be expanded before `foo`. Ordering is only guaranteed as far as it is
+necessary. E.g., if `bar` is only defined by expanding `foo`, then `foo` must be
+expanded before `bar`.
+
+## Function-like macro uses
+
+A function-like macro use (c.f., attribute-like macro use) is a macro use which
+uses `foo!(...)` or `foo! ident (...)` syntax (where `()` may also be `[]` or `{}`).
+
+Macros may be named by using a `::`-separated path. Naming follows the same
+rules as other items in Rust.
+
+If a macro `baz` (by example or procedural) is defined in a module `bar` which
+is nested in `foo`, then it may be used anywhere in the crate using an
+absolute path: `::foo::bar::baz!(...)`. It can be used via relative paths in the
+usual way, e.g., inside `foo` as `bar::baz!()`.
+
+Macros declared inside a function body can only be used inside that function
+body.
+
+For procedural macros, the path must point to the function defining the macro.
+
+The grammar for macros is changed, anywhere we currently parser `name "!"`, we
+now parse `path "!"`. I don't think this introduces any issues.
+
+Name lookup follows the same name resolution rules as other items. See [RFC
+1560](https://github.com/rust-lang/rfcs/pull/1560) for details on how name
+resolution could be adapted to support this.
+
+## Attribute-like macro uses
+
+Attribute macros may also be named using a `::`-separated path. Other than
+appearing in an attribute, these also follow the usual Rust naming rules.
+
+E.g., `#[::foo::bar::baz(...)]` and `#[bar::baz(...)]` are uses of absolute and
+relative paths, respectively.
+
+
+## Importing macros
+
+Importing macros is done using `use` in the same way as other items. An `!` is
+not necessary in an import item. Macros are imported into their own namespace
+and do not shadow or overlap items with the same name in the type or value
+namespaces.
+
+E.g., `use foo::bar::baz;` imports the macro `baz` from the module `::foo::bar`.
+Macro imports may be used in import lists (with other macro imports and with
+non-macro imports).
+
+Where a glob import (`use ...::*;`) imports names from a module including macro
+definitions, the names of those macros are also imported. E.g., `use
+foo::bar::*;` would import `baz` along with any other items in `foo::bar`.
+
+Where macros are defined in a separate crate, these are imported in the same way
+as other items by an `extern crate` item.
+
+No `#[macro_use]` or `#[macro_export]` annotations are required.
+
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+If the new macro system is not well adopted by users, we could be left with two
+very different schemes for naming macros depending on whether a macro is defined
+by example or procedurally. That would be inconsistent and annoying. However, I
+hope we can make the new macro system appealing enough and close enough to the
+existing system that migration is both desirable and easy.
```

Comment:
We could simply lint on `#[macro_use]` and `#[macro_export]`. This would never show up in dependencies, but only in the crate that's currently compiled